Add functionality to attach and detach StudySubject to Workshops

max-khakimzhanov opened the pull request:
This PR introduces **a complete set of features to attach and detach StudySubjects from Workshops.**
### **Key Changes:**  

### **🔹 Database Updates:**  
- Added a relationship between StudySubjects and Workshops.  
- Added a direct relationship between StudySubjects and Providers — now each StudySubject is explicitly linked to its owning provider.
- Created and applied a new migration to reflect these changes in the database schema.  

### **🔹 Services Updates**

**UpdateWorkshopsForStudySubject:**
- Implements logic to attach or detach workshops based on `IsAttached` flag.
- Validates **provider ownership** before applying changes.
- Logs cases when a passed workshop is not found.
- Optimized in-memory checks and LINQ calls for performance.

**DetachAllWorkshops:**
- Detaches all workshops belonging to a provider from a given study subject.
- Includes early exit checks to avoid unnecessary DB writes.
- Uses filtered queries to reduce loaded data.

**GetAttachedWorkshops:**
- Retrieves **paginated** list of provider’s workshops with their attachment status to a specific study subject.
- Applies efficient filtering and projection in a single query using LINQ.
- Prevents potential N+1 issues by ensuring eager loading where needed.

### **🔹 DTOs & Utilities**
- Introduced `WorkshopAttachmentStatusDto` for transmitting workshop ID, title, and attachment flag.
- Introduced `PaginatedResult<T>` and extension method `ToPaginatedResultAsync` for reusable pagination.

### **🔹 Controller & Routing**
 **RESTful Routes**:
- `GET /providers/{providerId}/studysubjects/{studySubjectId}/workshops/attached` — get paginated workshops with attachment status.
- `PUT /api/v1/providers/{providerId}/studysubjects/{studySubjectId}/workshops` — update (attach/detach) specific workshops.
- `DELETE /api/v1/providers/{providerId}/studysubjects/{studySubjectId}/workshops` — detach all workshops for a provider.

### **🔹Testing**
 **Unit tests added/updated** for:
- `GetAttachedWorkshops`: including pagination, filtering, and boundary checks.
- `UpdateWorkshopsForStudySubject`: checks ownership, attachment toggling, and invalid workshop handling.
- `DetachAllWorkshops`: verifies correct filtering and removal.
- Mocked `IQueryable` queries now support `ToPaginatedResultAsync` via `TestAsyncEnumerableQuery`.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Expanded workshop management by allowing study subjects to be linked with multiple workshops via provider associations.
  - Introduced endpoints to update, detach, and retrieve workshop attachment statuses with built-in pagination and sorting.
  - Enhanced API responses with improved validations and error handling for workshop management.

- **Tests**
  - Expanded test coverage to verify the new workshop management operations and pagination behavior, ensuring reliable functionality.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

max-khakimzhanov committed:
Add relationships for Provider, StudySubject, and Workshop.
Add fluent api configuration for this entities.

max-khakimzhanov committed:
Add relationships for Provider, StudySubject, and Workshop.
Add fluent api configuration for this entities.

max-khakimzhanov committed:
Add WorkshopAttachmentStatusDto and update mappings

- Refactor `StudySubjectDto`
- Updated `MappingProfile` to include new mappings for `StudySubjects`.

max-khakimzhanov committed:
Add WorkshopAttachmentStatusDto and update mappings

- Refactor `StudySubjectDto`
- Updated `MappingProfile` to include new mappings for `StudySubjects`.

max-khakimzhanov committed:
Add GetWorkshopsWithAttachmentStatusByProviderId method

- Implemented new method in IWorkshopService to retrieve workshops with attachment status for a given provider and study subject.
- Updated WorkshopService to include logic for the new method.
- Updated unit tests across multiple test classes to mock and pass the new repository to WorkshopService.

max-khakimzhanov committed:
Add GetWorkshopsWithAttachmentStatusByProviderId method

- Implemented new method in IWorkshopService to retrieve workshops with attachment status for a given provider and study subject.
- Updated WorkshopService to include logic for the new method.
- Updated unit tests across multiple test classes to mock and pass the new repository to WorkshopService.

max-khakimzhanov committed:
Add methods for attaching and detaching StudySubject to Workshops to StudySubjectService

- Introduced new methods in `IStudySubjectService` for updating and detaching workshops.
- Implemented the new methods in `StudySubjectService`.
- Modified unit tests in `StudySubjectServiceTests` to mock the new repository dependency.
- Adjusted `StudySubjectControllerTests` to accommodate changes.

max-khakimzhanov committed:
Improve workshop attachments methods in StudySubjectService

- Updated workshop retrieval to filter by provider ID.

max-khakimzhanov committed:
Add workshop management features to StudySubjectController

- Introduced `UpdateWorkshopAttachmentDto` for workshop ID management.
- Added methods for retrieving workshops, updating attachments, and detaching workshops.

max-khakimzhanov committed:
Add workshop management features to StudySubjectController

- Introduced `UpdateWorkshopAttachmentDto` for workshop ID management.
- Added methods for retrieving workshops, updating attachments, and detaching workshops.

max-khakimzhanov committed:
Refactor workshop attachment handling

- Updated `UpdateWorkshopAttachmentDto` to use `WorkshopAttachmentStatusDto` for managing workshop attachments and detachments.
- Added a new method in `IWorkshopServicesCombiner` to fetch workshops with their attachment status.
- Enhanced `UpdateWorkshopsForStudySubject` logic in `StudySubjectService` for better handling of workshop attachments
- Overall, these changes improve the clarity, functionality, and robustness of the workshop management processes.

max-khakimzhanov committed:
Refactor workshop attachment handling

- Updated `UpdateWorkshopAttachmentDto` to use `WorkshopAttachmentStatusDto` for managing workshop attachments and detachments.
- Added a new method in `IWorkshopServicesCombiner` to fetch workshops with their attachment status.
- Enhanced `UpdateWorkshopsForStudySubject` logic in `StudySubjectService` for better handling of workshop attachments
- Overall, these changes improve the clarity, functionality, and robustness of the workshop management processes.

max-khakimzhanov committed:
Refactor workshop-related tests and controller logic

- Added new tests for updating and detaching workshops in StudySubjectControllerTests.
- Introduced validation for null requests and empty IDs in WorkshopControllerTests.
- Updated StudySubjectController to change routing and remove IWorkshopService dependency.
- Modified UpdateWorkshopsForStudySubject to accept WorkshopAttachmentStatusDto.
- Changed DetachAllWorkshops to return NoContent on success.
- Enhanced documentation for methods in WorkshopController.

max-khakimzhanov committed:
Refactor workshop-related tests and controller logic

- Added new tests for updating and detaching workshops in StudySubjectControllerTests.
- Introduced validation for null requests and empty IDs in WorkshopControllerTests.
- Updated StudySubjectController to change routing and remove IWorkshopService dependency.
- Modified UpdateWorkshopsForStudySubject to accept WorkshopAttachmentStatusDto.
- Changed DetachAllWorkshops to return NoContent on success.
- Enhanced documentation for methods in WorkshopController.

max-khakimzhanov committed:
Add tests for retrieving and updating workshops, including error handling and attachment scenarios

max-khakimzhanov committed:
Update `Create` of StudySubjectService and refactor controllers

- Updated Create method of StudySubjectService to set ProviderId
- Simplified StudySubjectControllerTests by removing workshopServiceMock.
- Modified StudySubjectController to remove dependency on IWorkshopService.
- Added 204 No Content response type to GetWorkshopsWithAttachmentStatusByProviderId in WorkshopController.

max-khakimzhanov committed:
Update `Create` of StudySubjectService and refactor controllers

- Updated Create method of StudySubjectService to set ProviderId
- Simplified StudySubjectControllerTests by removing workshopServiceMock.
- Modified StudySubjectController to remove dependency on IWorkshopService.
- Added 204 No Content response type to GetWorkshopsWithAttachmentStatusByProviderId in WorkshopController.

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -94,7 +96,8 @@ public WorkshopService(
         IContactsService<Workshop, IHasContactsDto<Workshop>> contactsService,
-        IApplicationRepository applicationRepository)
+        IApplicationRepository applicationRepository,
+        IEntityRepositorySoftDeleted<Guid, StudySubject> studySubjectRepository)
```
We have special case for `GUID` keys - `ISensitiveEntityRepositorySoftDeleted`

max-khakimzhanov replied:
The `studySubjectRepository` is no longer used in this service — it was removed along with related logic during recent refactoring. I've already deleted it from the constructor.

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -356,6 +360,48 @@ public async Task<SearchResult<WorkshopProviderViewCard>> GetByProviderId(Guid i
+            .ConfigureAwait(false);
+
+        // Get the study subject with the specified Id and include the workshops
+        var studySubject = await studySubjectRepository
```
even with pagination, maybe a `left join` would be better? maybe as a little optional subtask try to test what runs faster.

Even if a bit slower - still will make the entire method fit into like 5 lines instead of making multiple array enumerations

max-khakimzhanov replied:
I've refactored the method to build WorkshopAttachmentStatusDto directly in the LINQ query.
This combines filtering, sorting, and checking attachment status in one step and avoids extra in-memory processing.

I also tested both versions — the updated one is faster.
The previous approach loaded unnecessary data and did more work in memory, while the new one runs as a single efficient SQL query

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/StudySubjectService.cs:
```diff
@@ -263,6 +272,150 @@ private async Task CheckIfLanguageIdIsCorrect(StudySubjectCreateUpdateDto dto)
+        foreach (var workshopDto in workshopsWithStatus)
+        {
+            var workshop = allWorkshops.FirstOrDefault(w => w.Id == workshopDto.Id);
+            if (workshop == null) continue; // Skip if workshop was not found
```
maybe log a warning or info that user passed non existing workshop into attachment.

max-khakimzhanov replied:
✅

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/StudySubjectService.cs:
```diff
@@ -263,6 +272,150 @@ private async Task CheckIfLanguageIdIsCorrect(StudySubjectCreateUpdateDto dto)
+    /// <typeparam name="T">The type of the entity that was not found.</typeparam>
+    /// <param name="id">The unique identifier of the missing entity.</param>
+    /// <returns>A failed <see cref="Result{T}"/> indicating that the entity was not found.</returns>
+    private Result<T> NotFoundResult<T>(Guid id)
```
it might be useful not only here, you can extract this helper method to Util

max-khakimzhanov replied:
✅

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Util/MappingProfile.cs:
```diff
@@ -863,10 +872,14 @@ public MappingProfile()
 
         CreateMap<StudySubject, StudySubjectDto>()
-            .ForMember(dest => dest.WorkshopId, opt => opt.Ignore());
+            .ForMember(dest => dest.Workshops, opt => opt.MapFrom(src => src.Workshops));
```
Haven't seen get method changes, but double check that the `src.Workshops` is populated by eager loading and not using N+1 query during mapping.

max-khakimzhanov replied:
I've checked this — `StudySubject.Workshops` is eager-loaded using `.GetByIdWithDetails(..., includeProperties: "Workshops")` before mapping.
So AutoMapper uses fully loaded data and no additional queries are made

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -356,6 +360,48 @@ public async Task<SearchResult<WorkshopProviderViewCard>> GetByProviderId(Guid i
+           Guid providerId)
+    {
+        // Get all workshops for the provider
+        var allWorkshops = await workshopRepository
```
There may be 100s of workshops per provider and workshops can have many subjects according to configuration. Maybe introduce pagination that will allow to fetch batches of data.

max-khakimzhanov replied:
Pagination has been added to the method and API endpoint as suggested.
The method now supports page and pageSize parameters, and returns a PaginatedResult<T> to avoid loading large datasets at once

DmyMi commented on OutOfSchool/OutOfSchool.DataAccess/Models/Configurations/ProviderConfiguration.cs:
```diff
@@ -89,5 +89,10 @@ public void Configure(EntityTypeBuilder<Provider> builder)
         builder.Property(x => x.UpdatedAt)
                 .ValueGeneratedOnAddOrUpdate();
+
+        builder.HasMany(p => p.StudySubjects)
```
is the duplicate configuration needed as the same is in `StudySubjectConfig`?

max-khakimzhanov replied:
Thanks for pointing that out! I've already removed the duplicate configuration

DmyMi commented on OutOfSchool/OutOfSchool.DataAccess/Models/Configurations/StudySubjectConfiguration.cs:
```diff
@@ -24,5 +24,13 @@ public override void Configure(EntityTypeBuilder<StudySubject> builder)
+        builder.HasMany(x => x.Workshops)
+            .WithMany(x => x.StudySubjects);
+
+        builder.HasOne(x => x.Provider)
```
check duplicate comments.

max-khakimzhanov replied:
✅

DmyMi commented on OutOfSchool/OutOfSchool.DataAccess/Models/Configurations/WorkshopConfiguration.cs:
```diff
@@ -40,5 +40,8 @@ public override void Configure(EntityTypeBuilder<Workshop> builder)
         builder.HasMany(x => x.Tags)
             .WithMany(x => x.Workshops);
+
+        builder.HasMany(x => x.StudySubjects)
```
is the duplicate configuration needed as the same is in `StudySubjectConfig`?

max-khakimzhanov replied:
This has already been deleted

DmyMi commented on OutOfSchool/OutOfSchool.DataAccess/Models/Provider.cs:
```diff
@@ -159,4 +159,5 @@ public class Provider : IKeyedEntity<Guid>, IImageDependentEntity<Provider>, ISo
     public DateTime UpdatedAt { get; set; }
 
     public virtual ICollection<Position> Positions { get; set; }
+    public virtual List<StudySubject> StudySubjects { get; set; }
```
I haven't seen any logic using provider as main point of interaction with subjects (i may be blind though 😅 ). So for now this navigation property is not needed

max-khakimzhanov replied:
You're right — there was no actual usage of that : ) 
I've removed it

DmyMi commented on OutOfSchool/OutOfSchool.DataAccess/Models/Workshop.cs:
```diff
@@ -152,8 +152,10 @@ public virtual ICollection<Workshop>
 
     public virtual List<Tag> Tags { get; set; }
+
+    public virtual List<StudySubject> StudySubjects { get; set; }
```
Same as with providers, i may be blind 😅 but most interaction goes from study subjects to workshops. So this may not be required

max-khakimzhanov replied:
Initially, it wasn’t used, but now this navigation property is needed. It’s used in WorkshopService.GetAttachedWorkshops to determine whether each workshop is attached to the given study subject

DmyMi replied:
As it is a join anyway (we need to take workshops and inner join to studysubjects) under the hood you can "reverse" it by taking the studysubjects and inner joining workshops.

And business logic wise this is more related to study subjects, so can be moved to studysubjects service (even the method args do not have workshops).

DmyMi commented on OutOfSchool/OutOfSchool.WebApi/Controllers/V1/WorkshopController.cs:
```diff
@@ -109,6 +109,44 @@ public async Task<IActionResult> GetWorkshopListByProviderId(Guid providerId)
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+    [HttpGet("{studySubjectId}/{providerId}")]
```
need to add something with path. As it should look like the following:
`/providers/{providerId}/studysubject/{studySubjectId}/workshops` (or `/workshops/attached` in the end).

Please think about the RESTful naming

max-khakimzhanov replied:
✅

DmyMi commented on OutOfSchool/OutOfSchool.WebApi/Controllers/V1/WorkshopController.cs:
```diff
@@ -109,6 +109,44 @@ public async Task<IActionResult> GetWorkshopListByProviderId(Guid providerId)
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+    [HttpGet("{studySubjectId}/{providerId}")]
+    public async Task<IActionResult> GetWorkshopsWithAttachmentStatusByProviderId(Guid studySubjectId, Guid providerId)
```
Think about adding pagination

max-khakimzhanov replied:
Pagination has been added to this endpoint.
It returns a PaginatedResult<WorkshopAttachmentStatusDto> and accepts page and pageSize as query parameters.
Also added the [ValidatePagination] attribute to validate the query parameters

DmyMi commented on OutOfSchool/OutOfSchool.WebApi/Controllers/V1/StudySubjectController.cs:
```diff
@@ -210,6 +210,71 @@ public async Task<IActionResult> Delete(Guid id, Guid providerId)
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(StatusCodes.Status403Forbidden)]
+    [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+    [HttpPut("{studySubjectId}")]
```
lets make new endpoints REST friendly :) think about RESTful naming for this and another new endpoint

max-khakimzhanov replied:
✅

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/StudySubjectService.cs:
```diff
@@ -263,6 +272,150 @@ private async Task CheckIfLanguageIdIsCorrect(StudySubjectCreateUpdateDto dto)
+        Guid providerId,
+        IEnumerable<WorkshopAttachmentStatusDto> workshopsWithStatus)
+    {
+        await providerService.HasProviderRights(providerId).ConfigureAwait(false);
```
Also, do not use `providerService.HasProviderRights`, this method will be removed. Use `ICurrentUserService` directly

max-khakimzhanov replied:
✅

max-khakimzhanov committed:
Remove duplicate configurations, unnecessary navigation property and refactor mapping

max-khakimzhanov committed:
Refactor workshop service for pagination support

Updated `IWorkshopService` to replace `GetWorkshopsWithAttachmentStatusByProviderId` with `GetAttachedWorkshops`, adding pagination parameters.
Modified `WorkshopService` to implement the new method and adjusted logic accordingly.
Introduced `PaginationExtensions` for async pagination and added `PaginatedResult<T>` class to encapsulate paginated query results.
Updated related test classes to reflect these changes and ensure proper testing of the new functionality.

max-khakimzhanov committed:
Refactor StudySubjectService and enhance tests

- Replaced `providerService.HasProviderRights` with `currentUserService.UserHasRights` in `UpdateWorkshopsForStudySubject`.
- Moved `NotFoundResult` method to `OperationResultHelper` for reuse.
- Added unit tests in `PaginationExtensionsTests.cs` for pagination functionality.
- Updated `StudySubjectServiceTests.cs` to align with service method changes.
- Enhanced `TestAsyncEnumerableQuery` for better async operation support.

max-khakimzhanov committed:
Refactor attached workshop-related endpoints and routes to make them RESTfull

Updated tests for `WorkshopController`

max-khakimzhanov committed:
Add ValidatePaginationAttribute for API pagination checks

Introduces a new `ValidatePaginationAttribute` class that validates the `page` and `pageSize` parameters in API methods. Ensures both values are greater than 0, returning a `BadRequestObjectResult` with an error message if validation fails.

max-khakimzhanov committed:
Reapply state of 3378f75dea56ec8d63a8b155307690a049b0d8f2

max-khakimzhanov committed:
Fix rebase issues

max-khakimzhanov commented on OutOfSchool/OutOfSchool.WebApi/Util/ValidatePaginationAttribute.cs:
```diff
@@ -0,0 +1,28 @@
+            pageSizeObj is int pageSize && pageSize < 1)
+        {
+            context.Result = new BadRequestObjectResult("Page size must be greater than 0.");
+            return;
```
this redundant return will be deleted

DmyMi commented on OutOfSchool/OutOfSchool.Migrations/Data/Migrations/OutOfSchoolMigrations/20250409143340_AddStudySubjectRelations.cs:
```diff
@@ -0,0 +1,83 @@
+                table: "StudySubjects",
+                type: "UUID(16)",
+                nullable: false,
+                defaultValue: new Guid("00000000-0000-0000-0000-000000000000"));
```
it is a `NOT NULL` field that should never have a default property. I have not seen any migration before creating a default value for a foreign key. Yes it would fail a foreign key constraint most of the time, but there might be an error and some one adds a provider with empty Guid and this default value will bind to that provider

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Services/Database/WorkshopService.cs:
```diff
@@ -335,6 +336,30 @@ public async Task<SearchResult<WorkshopProviderViewCard>> GetByProviderId(Guid i
+           int page,
+           int pageSize)
+    {
+        logger.LogInformation("Getting workshops with attachment status. ProviderId = {ProviderId}, " +
```
This log should be `Debug`

max-khakimzhanov committed:
Add migration, refactor `StudySubjectConfiguration` and fix issue

- fix issue - remove reduredundant redundant  in `ValidatePaginationAttribute`
